File: server/src/algo_py/improve.py
import copy
import sys
import json
from util import order_metric, overall_metric, rotation, perturbation
from box import Box
from container import Container

NUMBER_OF_ITERATIONS = 1000


def improve_packing(in_boxes: list[Box], out_boxes: list[Box], container: Container) -> list[Box]:

    # We are receiving a list boxes with isIn bit on or off.
    # We need to treat boxes with isIn bit on as constants.
    # We need to find pps from this boxes.
    # After having those pps, we need to run the algorithm with those pps and the rest of the boxes.

    # Updating pps by every box

    pp = set([(0, 0, 0, 1), (container.size[0] - 1, 0, 0, -1)])

    retry_list = []
    solution_boxes = []
    solution_data = {"number_of_items": 0, "capacity": 0,
                     "order_score": 0, "overall_score": 0}

    # Finding pps from in_boxes.
    for box in in_boxes:
        # Placing the box in it right position
        p_left = (int(box.FLB[0]), int(box.FLB[1]), int(box.FLB[2]), 1)
        container.place(box, p_left)
        p_right = (box.FLB[0], box.FLB[1], box.FLB[2], -1)
        # Updating all possible pps from this box
        pp.add(p_left)
        container.update(box, p_left, pp)
        pp.add(p_right)
        container.update(box, p_right, pp)
        solution_boxes.append(box)
        solution_data['number_of_items'] += 1
        solution_data['capacity'] += box.volume

    # Adding each boxes in out_boxes to the container
    for box in out_boxes:
        best_point = None
        best_score = (0, 0)
        # to find the best point we need to know what corner of the box is placed there
        for p in pp:
            score = container.get_score(box, p)
            if score > best_score:
                best_point = p
                best_score = score
        if best_point:
            container.place(box, best_point)
            container.update(box, best_point, pp)
            solution_boxes.append(box)
            solution_data['number_of_items'] += 1
            solution_data['capacity'] += box.volume
        else:
            retry_list.append(box)

    for box in retry_list:
        best_point = None
        best_score = (0, 0)
        for p in pp:
            score = container.get_score(box, p)
            if score > best_score:
                best_point = p
                best_score = score
        if best_point:
            container.place(box, best_point)
            container.update(box, best_point, pp)
            solution_boxes.append(box)
            solution_data['number_of_items'] += 1
            solution_data['capacity'] += box.volume
        else:
            # Adding it to the list without adding it to the container
            solution_boxes.append(box)

    boxes = in_boxes + out_boxes
    solution_data['order_score'] = order_metric(
        solution_boxes, boxes, container)
    solution_data['overall_score'] = overall_metric(
        boxes, container, solution_data, 1)
    return solution_boxes, solution_data


def improve():

    # every key in json is a string in python dict.
    obj = json.loads(sys.argv[1])

    

    container = Container(obj['container']['width'],
                          obj['container']['height'],
                          obj['container']['length'])

    boxes = []

   

    for b in obj['boxes']:
        boxes.append(Box(b['id'], b['order'], b['text'], b['size'][0],
                         b['size'][1], b['size'][2], color=b['color'], isIn=b['isIn'], center=b['position']))

     # Splitting to two lists
    in_boxes, out_boxes = [box for box in boxes if box.isIn == 1], [
        box for box in boxes if box.isIn == 0]
    # Sort isIn's by y coordinate and z coordinate
    in_boxes = sorted(in_boxes, key=lambda box: (box.FLB[1], box.FLB[2]))

    solution_list = {}
    counter = 0

    


    for _ in range(NUMBER_OF_ITERATIONS):
        copy_in_boxes = copy.deepcopy(in_boxes)
        copy_out_boxes = copy.deepcopy(out_boxes)
        container.start_packing()

        # For non-deterministic algorithm
        rotation(copy_out_boxes)
        perturbation(copy_out_boxes)

        temp = improve_packing(copy_in_boxes, copy_out_boxes, container)
        if temp is None:
            continue

        boxes_in_solution, solution_data = temp
        if boxes_in_solution is not None and solution_data is not None:
            counter_string = f'{counter}'
            solution_list[counter_string] = {
                "name": "solution " + counter_string, "id": counter, "boxes": boxes_in_solution, "solution_data": solution_data}
            counter += 1

    solution_list = sorted(solution_list.values(
    ), key=lambda x: x["solution_data"]["overall_score"], reverse=True)[0]

    return solution_list


print(improve())

File: server/src/algo_js/improve.js
const { Rotation } = require("./boxMotion.js");
const { getSize } = require("./box.js");
const { updatePps } = require("./container.js");
const { rotation, perturbation } = require("./boxMotion.js");
const { orderMetric, overallMetric } = require("./metrics.js");
const { handleBox } = require("./algo.js");

const IMPROVE_TIME = 10000;
const IMPROVE_MAX_ITERATIONS = 1000000;

const improvePacking = (inBoxes, outBoxes, container) => {
	let pp = new Set([
		{
			x: 0,
			y: 0,
			z: 0,
			dir: 1,
		},
		{
			x: container.width,
			y: 0,
			z: 0,
			dir: -1,
		},
	]);

	let solutionBoxes = [];
	let retryList = [];

	let solution_data = {
		number_of_items: 0,
		capacity: 0,
		order_score: 0,
		overall_score: 0,
	};

	// Finding pps from inBoxes.
	inBoxes.forEach((box) => {
		let pLeft = {
			...box.FLB,
			dir: 1,
		};
		let pRight = {
			...box.FLB,
			dir: -1,
		};
		// Updating all possible pps from this box
		pp.add(pLeft);
		updatePps(box, pLeft, pp, container);
		pp.add(pRight);
		updatePps(box, pRight, pp, container);
		solutionBoxes.push(box);
		solution_data["number_of_items"] += 1;
		solution_data["capacity"] += box.volume;
	});

	// Adding each box in outBoxes to the container
	outBoxes.forEach((box) => {
		handleBox(
			box,
			pp,
			container,
			solutionBoxes,
			solution_data,
			retryList,
			false
		);
	});

	retryList.forEach((box) => {
		handleBox(
			box,
			pp,
			container,
			solutionBoxes,
			solution_data,
			retryList,
			true
		);
	});

	const boxes = [...inBoxes, ...outBoxes];
	solution_data.order_score = parseFloat(
		orderMetric(solutionBoxes, container)
	);
	solution_data.overall_score = parseFloat(
		overallMetric(boxes, container, solution_data, 1)
	);
	return [solutionBoxes, solution_data];
};

const initImproveBox = (box) => {
	return {
		...box,
		width: box.size[0],
		height: box.size[1],
		length: box.size[2],
		volume: box.size[0] * box.size[1] * box.size[2],
		rotation: Rotation.WHL,
		size: {
			width: box.size[0],
			height: box.size[1],
			length: box.size[2],
		},
		center: {
			x: box.position[0],
			y: box.position[1],
			z: box.position[2],
		},
		FLB: {
			x: box.position[0] - 0.5 * box.size[0],
			y: box.position[1] - 0.5 * box.size[1],
			z: box.position[2] - 0.5 * box.size[2],
		},
	};
};

// Returns the box with the correct size.
const getImproveBox = (box) => {
	return {
		...box,
		size: Object.values(getSize(box)),
		position: Object.values(box.center),
	};
};

const improve = (data) => {
	console.log(data);

	const container = data.container;
	let boxes = data.boxes;

	boxes = boxes.map((box) => {
		return initImproveBox(box);
	});

	// Splitting to two lists
	const inBoxes = boxes.filter((box) => {
		return box.isIn == 1;
	});
	const outBoxes = boxes.filter((box) => {
		return box.isIn == 0;
	});

	// Sort isIn's by y coordinate and z coordinate
	inBoxes.sort((box1, box2) => {
		if (box1.FLB.y === box2.FLB.y) {
			return box1.FLB.z - box2.FLB.z;
		}
		return box1.FLB.y - box2.FLB.y;
	});

	let solutionList = {};
	let counter = 0;

	const endTime = Date.now() + IMPROVE_TIME;
	while (Date.now() < endTime) {
		inBoxesCopy = [...inBoxes];
		outBoxesCopy = [...outBoxes];

		outBoxesCopy = rotation(outBoxesCopy);
		outBoxesCopy = perturbation(outBoxesCopy);

		let solution = improvePacking(inBoxesCopy, outBoxesCopy, container);
		if (solution === null) {
			continue;
		}

		let [boxesInSolution, solution_data] = solution;
		if (boxesInSolution !== null && solution_data !== null) {
			let counterString = counter.toString();
			solutionList[counterString] = {
				name: "solution " + counterString,
				id: counter,
				boxes: boxesInSolution.map((box) => {
					return getImproveBox(box);
				}),
				solution_data: solution_data,
			};
			counter += 1;
		}
		if (counter === IMPROVE_MAX_ITERATIONS) {
			break;
		}
	}

	// The best solution is the one with the most boxes in it.
	const bestSolution = Object.values(solutionList).sort(
		(a, b) =>
			b.solution_data.number_of_items - a.solution_data.number_of_items
	)[0];

	return bestSolution;
};

module.exports = { improve };
